### PR TITLE
add matrix to QubitStateVector operation

### PR DESCRIPTION
**Pull request by timmysilv**
This is the same as #3633, but for QubitStateVector

**Context:**
Prep operations should have a matrix representation that is a ket vector.

**Description of the Change:**
Adds `matrix()` and `has_matrix` to `qml.QubitStateVector`

**Benefits:**
The `QubitStateVector` operation can be used in matrix multiplication during computations. This is primarily for use by simulators, and should be useful for any simulator.

**Related GitHub Issues:**
I first attempted something to a similar effect in #3626, but this approach makes more sense.

### CODE REVIEW

File: pennylane/ops/qubit/state_preparation.py [hunk 1]
```diff
@@ -156,6 +158,32 @@ def compute_decomposition(state, wires):
...
+        return True
+
+    def matrix(self, wire_order=None):
+        """Returns a ket vector representing the state being created.
+
+        QSV([1/2, 1/2, 1/2, 1/2], wires=[0,2]) = 1/4 * (|0 0> + |0 1> + |1 0> + |1 1>)
+                                               = 1/4 * (|000> + |001> + |100> + |101>)
+        """
+        if wire_order is None:
+            return np.array(self.parameters[0], dtype="complex128")
```
Comment (josh146) [inline pennylane/ops/qubit/state_preparation.py:173]: @timmysilv how come you are casting to NumPy here? I am thinking of a use case where someone has

```python
QubitStateVector(torch.tensor(...), wires=...)
```

File: pennylane/ops/qubit/state_preparation.py [hunk 2]
```diff
@@ -156,6 +158,32 @@ def compute_decomposition(state, wires):
...
+        if not wires.contains_wires(self.wires):
+            raise WireError("Custom wire_order must contain all QubitStateVector wires")
+        num_wires = len(wires)
+        indices = [0] * num_wires
+        for base_wire_idx in [wires.index(w) for w in self.wires]:
+            indices[base_wire_idx] = slice(None)  # same as ":"
+
+        ket = np.zeros((2,) * num_wires)
+        ket[tuple(indices)] = np.reshape(self.parameters[0], (2,) * len(self.wires))
+        return ket.reshape(2**num_wires)
```
Comment (josh146) [inline pennylane/ops/qubit/state_preparation.py:185]: @timmysilv I wonder if you can re-use either https://github.com/PennyLaneAI/pennylane/blob/2c595a67d541dab4da8461ce8b6336fe25ad298e/pennylane/math/matrix_manipulation.py#L27 or https://github.com/PennyLaneAI/pennylane/blob/2c595a67d541dab4da8461ce8b6336fe25ad298e/pennylane/utils.py#L252 here?

Comment (josh146) [reply pennylane/ops/qubit/state_preparation.py:185]: (`expand_matrix` ideally since it is newer, and `expand_vector` I think is a bit forgotten)

**Comment by trbromley** [comment]
Have we resolved @mariaschuld's concerns about whether `.matrix` should return the state vector or a unitary what prepares the state vector?

**Comment by mariaschuld** [comment]
> Have we resolved @mariaschuld's concerns about whether `.matrix` should return the state vector or a unitary what prepares the state vector?

Yes, I got convinced that for a lot of users this would be intuitive, which may outweigh that it is not consistent (since `QubitStateVector` is an operator, not a state). No objections :)

File: pennylane/ops/qubit/state_preparation.py (hunk 1 above)
Comment (timmysilv) [reply pennylane/ops/qubit/state_preparation.py:173]: the goal was more to cast to complex128. My original plan was for this to be consumed by a particular state-prep procedure with a `like=some_interface` argument, but it seems like that won't be all encompassing so I'll support the preservation of interfaces here (as I'm doing in the linked BasisState PR now as well)

File: pennylane/ops/qubit/state_preparation.py (hunk 2 above)
Comment (josh146) [reply pennylane/ops/qubit/state_preparation.py:185]: nice!

Comment (josh146) [reply pennylane/ops/qubit/state_preparation.py:185]: A good time to unify `expand_matrix`, `expand_vector`, and the new one where possible.

Comment (timmysilv) [reply pennylane/ops/qubit/state_preparation.py:185]: I'll take note of it, see how we can bring it all together outside of this PR

**Comment by timmysilv** [comment]
the [latest CI errors](https://github.com/PennyLaneAI/pennylane/actions/runs/3932991010/jobs/6726201491) are making me re-think this. We could make `adjoint(op).matrix()` work, but does it have eigvals defined? is the adjoint just the conjugate (numpy does nothing if you transpose a vector)? this is only selectively valuable and I'm becoming unsure. Whatever we decide, I'll add tests to confirm it. I was hoping to push off adjoint support but `test_adjoint` singles this class out! lesson learned.

**Comment by timmysilv** [comment]
I just pushed a commit to add adjoint support for the operator. I feel unsure about whether or not this was the correct approach, mostly because I added a (small) conditional check to _every_ invocation of `Adjoint.matrix` regardless of the base class. Feedback on that commit would be greatly appreciated 😄

**Comment by timmysilv** [comment]
see linked PR instead

**Pull request closed without merging**